### PR TITLE
Deal with explicit property deletions inside abstract loops

### PR DESCRIPTION
**Pull request by hermanventer**
Release note: none

Add some test cases that have loops that delete properties in some iterations but not others.
Deal with it by explicitly testing for __empty values.

Issue #1174

### CODE REVIEW

File: src/methods/widen.js [hunk 1]
```diff
@@ -297,7 +297,7 @@ export class WidenImplementation {
       if (val1 === undefined) continue; // deleted
       let val2 = m2.get(key1);
       if (val2 === undefined) continue; // A key that disappears has been widened away into the unknown key
-      if (!f(val2, val1)) return false;
+      if (!f(val1, val2)) return false;
```
Comment (hermanventer) [inline src/methods/widen.js]: Fixing a typo that only manifested as a bug in the new test cases.

File: src/values/AbstractValue.js [hunk 2]
```diff
@@ -245,31 +245,31 @@ export default class AbstractValue extends Value {
   mightBeNull(): boolean {
     let valueType = this.getType();
     if (valueType === NullValue) return true;
-    if (valueType !== Value) return false;
+    if (valueType !== PrimitiveValue && valueType !== Value) return false;
```
Comment (hermanventer) [inline src/values/AbstractValue.js]: Until widening came in to the picture, no test case caused an abstract value with type === PrimitiveValue.

File: src/values/AbstractValue.js [hunk 3]
```diff
@@ -245,31 +245,31 @@ export default class AbstractValue extends Value {
...
     if (this.values.isTop()) return true;
     return this.values.includesValueOfType(NumberValue);
   }
 
   mightNotBeNumber(): boolean {
     let valueType = this.getType();
-    if (valueType === NumberValue) return false;
-    if (valueType !== Value) return true;
+    if (Value.isTypeCompatibleWith(valueType, NumberValue)) return false;
```
Comment (hermanventer) [inline src/values/AbstractValue.js]: If we ever introduce an IntegralValue as a subtype of NumberValue, this kind of thing will be necessary, so may as well do it now while touching this code.

File: src/realm.js [hunk 4]
```diff
@@ -583,35 +584,63 @@ export class Realm {
...
-          let inTest = t.binaryExpression("in", t.stringLiteral(key.key), o);
-          let vDefined = t.binaryExpression("!==", v, t.unaryExpression("void", t.numericLiteral(0)));
-          let guard = t.logicalExpression("||", vDefined, inTest);
-          return t.ifStatement(guard, r);
+          // If v === __empty || (v === undefined  && !(key.key in o))  then delete it
+          let emptyTest = t.binaryExpression("===", v, e);
+          let undefinedTest = t.binaryExpression("===", v, voidExpression);
+          let inTest = t.unaryExpression("!", t.binaryExpression("in", t.stringLiteral(key.key), o));
+          let guard = t.logicalExpression("||", emptyTest, t.logicalExpression("&&", undefinedTest, inTest));
+          let deleteIt = t.expressionStatement(t.unaryExpression("delete", (lh: any)));
```
Comment (cblappert) [inline src/realm.js]: Could you add an invariant that lh is the right type?

Comment (hermanventer) [reply src/realm.js]: Sadly no. The expected type is a Flow union and there is no way I know of to assert this dynamically.

**Pull request closed without merging**